engine-src/Game/LambdaHack/Client/UI/HandleHumanM.hs

```diff
@@ -56,21 +58,35 @@ noRemoteHumanCmd cmd = case cmd of
   ContinueToXhair -> True
   _ -> False
 
+updateLastAction :: K.KM -> HumanCmd -> [ActionBuffer] -> [ActionBuffer]
+updateLastAction km cmd abuffs = case cmd of
+  RepeatLast{} -> abuffs
+  _ -> let oldBuffer = head abuffs
+           newBuffer = oldBuffer { slastAction = Just km }
+       in newBuffer : tail abuffs
+
 cmdAction :: (MonadClient m, MonadClientUI m)
-          => HumanCmd -> m (Either MError ReqUI)
-cmdAction cmd = case cmd of
+          => K.KM -> HumanCmd -> m (Either MError ReqUI)
+cmdAction km cmd = do
+  modifySession $ \sess ->
+    sess {sactionPending = updateLastAction km cmd $ sactionPending sess}
+  cmdActionCases cmd
```

Comment:
Point taken about naming. The name is very old and no longer applicable. 

Do you mean product type, which is a record or tuple? A function taking two argument is isomorphic to a functinon a tuple, so that figures. However, function `cmdActionCases`, which I will rename to `cmdSemantics` still only takes a command, so I don't think we should merge commands and keys permanently.

Reply:
And I'd rename `cmdAction` to `cmdSemInCxtOfKM`. which means semantics of the command in context of the given km as the last action.

Reply:
Perfect. `cmdSemInCxtOfKM` is gross, but does the job, since it's more like a helper function than actual handler. The thing is I wasn't positive about checking if action can be saved as the last action should be a part of any action semantics.

Reply:
Heh, `cmdSemInCxtOfKM` is gross. :D 

Yes, checking keystrokes is part of the semantics of some commands, namely of the commands that open menus that list keys and commands, which you can then pick. The semantics of such commands is recursive and involves not only looking up what command a key invokes, but also adding the key as the last action, etc. These are: help, dashboard, main menu, item manipulation menus and others.